core/src/main/java/com/radixdlt/sync/LocalSyncService.java

```diff
@@ -428,68 +418,137 @@ private boolean isFullySynced(SyncState.SyncingState syncingState) {
-
-    if (!currentState.waitingForResponseFrom(sender)) {
-      log.trace("LocalSync: Received unexpected sync response from {}", sender);
-      metrics.sync().unexpectedSyncResponses().inc();
-      return currentState;
-    }
-
-    final var responseTailStateVersion =
-        syncResponse.getLedgerExtension().getTail().getLedgerHeader().getStateVersion();
-    final var currentStateVersion = currentState.getCurrentHeader().getStateVersion();
-    if (responseTailStateVersion <= currentStateVersion) {
-      log.trace("LocalSync: Received a stale sync response from {}", sender);
-      metrics.sync().staleSyncResponses().inc();
-      return this.processSync(currentState.clearPendingRequest());
-    }
-
-    if (syncResponse.getLedgerExtension().getTransactions().isEmpty()) {
-      log.warn("LocalSync: Received empty sync response from {}", sender);
-      // didn't receive any transactions, remove from candidate peers and processSync
-      return this.processSync(currentState.clearPendingRequest().removeCandidate(sender));
-    }
-
-    if (!this.verifyConsensusProofOnSyncResponse(syncResponse)) {
-      log.warn(
-          "LocalSync: Received consensus-mismatched sync response {} from {}",
-          syncResponse,
-          sender);
-      // consensus-level validation failed, remove from candidate peers and processSync
-      invalidSyncResponseHandler.handleInvalidSyncResponse(sender, syncResponse);
-      return this.processSync(currentState.clearPendingRequest().removeCandidate(sender));
-    }
-
     try {
-      this.verifiedSyncResponseHandler.handleSyncResponse(sender, syncResponse);
-    } catch (InvalidCommitRequestException exception) {
-      log.warn(
-          "LocalSync: Received invalid commit request ({}) in sync response {} from {}",
-          exception.getMessage(),
-          syncResponse,
-          sender);
-      // ledger-level validation failed, remove from candidate peers and processSync
-      invalidSyncResponseHandler.handleInvalidSyncResponse(sender, syncResponse);
-      return this.processSync(currentState.clearPendingRequest().removeCandidate(sender));
+      this.syncResponseHandler.handle(currentState, sender, syncResponse);
+      // At this point we know that the commit was successful:
+      this.metrics.sync().validResponsesReceived().inc();
+      this.peerControl.reportHighPriorityPeer(sender);
+      this.syncLedgerUpdateTimeoutDispatcher.dispatch(
+          SyncLedgerUpdateTimeout.create(currentState.getCurrentHeader().getStateVersion()), 1000L);
+      return currentState.clearPendingRequest();
+    } catch (InvalidSyncResponseException isre) {
+      // Implementation note:
+      // Java 17 still cannot compile a "complete catch" of all subclasses of a sealed exception.
+      // But it can compile a "complete switch", and we want to leverage this in order to ensure
+      // that all future corner-cases of response processing are handled here as well.
+      return switch (isre) {
```

Comment:
_(note from author)_

As you can see below, I went for a rich and explicit exception hierarchy to handle different outcomes of sync response processing.
The `LocalSyncService` cares about maintaining the sync state machine and its infra (e.g. rewarding/penalizing peers, bumping metrics).
The extracted `SyncResponseHandler` cares only about validating and committing the transactions from the response.

You may ask "why validation vs commit cannot be separate responsibilities?".
That's because a critical part of validation (e.g. transaction parsing, ledger hashes verification) is done by a downstream (Rust) state manager and would be rather hard to pull out from there. So I kinda treat "commit or throw if anything was invalid" as a single operation, which can either return nothing (== success) or throw some very specific exception (so that the caller can precisely drive the sync state's machine reaction).

You may also ask "do we need exceptions? can't we return something like `SyncResponseHandlingResult` enum which summarizes the outcome?".
Currently, for technical reasons (i.e. the `EventDispatcher<>` interface, which we use everywhere, even for 100% sync calls), it wouldn't be so easy (we would need a refactor: migrate the "committer" to a plain sync service).
I also considered a hardcore checked `Exception` (since in a BFT world, invalid response is more of a "contingency" than "unexpected error"), but I know many people hate them.